### PR TITLE
Fix #4013 _resolve_host race conditions

### LINKED ISSUE #4013: Potential cache/throttle race conditions in Connector._resolve_host with traces
## Long story short

If `traces` are used in [`TCPConnector._resolve_host`](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L754) the context switch introduced by awaiting their trace API calls can lead to various race conditions.

For example the [DNS cache hit test](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L779-L780) may evaluate `True` but by the time `traces` is [done](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L786) it may already be gone again.

The same is true for the throttling logic where between checking that the [key is already throttled](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L788) and then [accessing that key](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L792) the key may already be popped.

Similarly in the [opposite case](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L793) of the above, with traces enabled, the code might [override an already set throttled event](https://github.com/aio-libs/aiohttp/blob/b35c9e3facaa50093feb05309fb7e679ffbb248b/aiohttp/connector.py#L797-L798)

## Steps to reproduce

See [my branch](https://github.com/ojii/aiohttp/blob/93dd3657a667d6367848ee0bb45bbab2ac25472d/tests/test_connector.py#L2263-L2289)

In the tests I mess with the caches from the traces directly to simulate the effects for simplicity.

### PR DESCRIPTION
**Pull request by ojii**
## What do these changes do?

Fixes #4013 by removing possible context switches between dictionary/cache tests and accessing those dictionaries/caches.

## Related issue number

Fixes #4013

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by webknjaz** [comment]
@ojii this PR causes linter failures. Please fix those and go through the checklist above :point_up:

**Comment by webknjaz** [comment]
Could you please also rebase this branch after sorting the imports?

**Comment by webknjaz** [comment]
I won't review this PR, plz handle the merge/backport this w/o me.

**Comment by dimaqq** [comment]
👍 
I wish the new tests made it clearer what was being tested, for example `DummyTracer` is hardly a good name for "tracer with a side effect that simulates another coro calling resolver which may change shared data", but I guess that's minor...

File: aiohttp/connector.py [hunk 1]
```diff
@@ -778,24 +778,25 @@ def clear_dns_cache(self,
 
         if (key in self._cached_hosts) and \
                 (not self._cached_hosts.expired(key)):
+            result = self._cached_hosts.next_addrs(key)
```
Comment (asvetlov) [inline aiohttp/connector.py:782]:
```suggestion
            # get result early, before any await (#4014)
            result = self._cached_hosts.next_addrs(key)
```

File: aiohttp/connector.py [hunk 2]
```diff
@@ -778,24 +778,25 @@ def clear_dns_cache(self,
...
 
             if traces:
                 for trace in traces:
                     await trace.send_dns_cache_hit(host)
-
-            return self._cached_hosts.next_addrs(key)
+            return result
 
         if key in self._throttle_dns_events:
+            event = self._throttle_dns_events[key]
```
Comment (asvetlov) [inline aiohttp/connector.py:791]:
```suggestion
            # get event early, before any await (#4014)
            event = self._throttle_dns_events[key]
```

File: aiohttp/connector.py [hunk 3]
```diff
@@ -778,24 +778,25 @@ def clear_dns_cache(self,
...
             if traces:
                 for trace in traces:
                     await trace.send_dns_cache_hit(host)
-            await self._throttle_dns_events[key].wait()
+            await event.wait()
         else:
+            self._throttle_dns_events[key] = \
```
Comment (asvetlov) [inline aiohttp/connector.py:798]:
```suggestion
            # update dict early, before any await (#4014)
            self._throttle_dns_events[key] = \
```

**Comment by asvetlov** [comment]
Regarding @dimaqq note I can live with the current naming but if you guys want to use more meaningful names -- you are welcome.

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**